### ISSUE TITLE
Allow tests to reference the key in messages to users

### ISSUE THREAD

**Issue by savage-alex** [stoplightio/spectral#1339]
As a rule creator (and a consumer of the report)
I want to receive verbose messages informing me of the location of an issue with its name
So that I can quickly see what went wrong with my design.

(This makes allot more sense when designing on an unresolved definition and exporting a resolved one into spectral.)

Given the following rule:
![image](https://user-images.githubusercontent.com/25430683/93092330-34875380-f697-11ea-8d35-2ed5a71ced0e.png)
The consumer is told exactly which parameter has the fault (and the line and ruleset) 
![image](https://user-images.githubusercontent.com/25430683/93092054-d9edf780-f696-11ea-92dc-63b071ade674.png)
Perfect right?

Unfortunately with truthy evaluations for a property it doesn't seem possible to get the key out.
I have tried to include the"hint" {parameter} or {value} in the rule:

Rule using property to pull through:
![image](https://user-images.githubusercontent.com/25430683/93091759-71067f80-f696-11ea-9dae-10d072d503ba.png)
Results in:
![image](https://user-images.githubusercontent.com/25430683/93091818-867ba980-f696-11ea-870c-55c987ec28af.png)

Rule changed to return the value, results in:

![image](https://user-images.githubusercontent.com/25430683/93091676-5207ed80-f696-11ea-88cf-e2994cd36d13.png)

Is it possible we can get the key returned as an option? I think it would be really helpful to consumers

**Comment by P0lip** [comment]
Hey!
Key should be already supported.
Did you try using `{property}`?

**Comment by savage-alex** [comment]
@P0lip hey there, see the above extract when using {{property}} i get a response back of [1] example

Can provide you the API definition and rule if it helps?

**Comment by P0lip** [comment]
Ah, it was said `{parameter}` in the initial description, so I assumed you simply misspelled it, but now I see it returns `[1].example`.
If I understand the issue correctly, you'd like to have just `example` printed out, right?
If not, what would be the ideal output for you?

**Comment by savage-alex** [comment]
It would be great if it were able to provide the value from another property? such as parameter name?
Given the example below
![image](https://user-images.githubusercontent.com/25430683/93198424-ec266f00-f744-11ea-8a96-d6c0ca5328c1.png)
It would be super useful to report the name of the parameter at fault.

**Comment by P0lip** [comment]
Agreed. That'd indeed provide a much more meaningful feedback.
I'll try to see what we can do about it.

**Comment by savage-alex** [comment]
Trying to get to the point that @arno-di-loreto has with the path / operation tests surfacing really nice responses ala https://apihandyman.io/toolbox/spectral/
![image](https://user-images.githubusercontent.com/25430683/93208105-eedc9080-f753-11ea-9e3f-2cfe8ad9e9b6.png)

I have the rules for 400, 401 .... 415 when there is a request body.. But would really level up with the path and operation in the message

**Comment by arno-di-loreto** [comment]
If the question is “how do I get the path in message “, I use the following conf using path variable on each rule:

message: “{{description}} ({{path}})”

**Comment by savage-alex** [comment]
Way to save the day and give me loads of JEST tests to change ;-) 
![image](https://user-images.githubusercontent.com/25430683/93210156-06694880-f757-11ea-8869-408149ecb9e8.png)
Thank you @arno-di-loreto ! Solves one on my problems. What do you think about the top request?

**Comment by P0lip** [comment]
@savage-alex I put some draft PR over here https://github.com/stoplightio/spectral/pull/1341, which should address your first concern. I'm not quite sure as to its final shape yet, but either way, please let me know whether it meets your requirements.

**Comment by savage-alex** [comment]
Hey @P0lip, im afraid its a bit beyond me to review. Happy to have a quick MS teams call if your game to take a look at it in the flesh?

**Comment by P0lip** [comment]
Hey, no worries, you don't need to review it.
I'd just like you to take a look whether the output you desire would be possible to generate once my change is merged.
I can compile a Spectral binary if you want to try it out and don't want to clone the whole Spectral repo. 
Does that sound good?

**Comment by savage-alex** [comment]
Hey, I have the PR running locally. Given the following OAS parameter against an operation:
 parameters:
        - in: query
          name: includeInactive
          description: false by default. If true, includes inactive tenants in the response
          schema:
            type: boolean
          required: false

there is no example. My rule currently sits at:

adv-query-parameters-example-exists:
    description: All query parameters SHOULD have an example.
    severity: error
    message: 'All query parameters SHOULD have an example.'
    given: "$..parameters[?(@.in == 'query')]"
    then:
      field: example
      function: truthy

To retrieve the property name:includeInactive in the message as that is where the fault is, what would the message need to have?

**Comment by P0lip** [comment]
@savage-alex 
Hey, this would be
```yaml
rules:
  adv-query-parameters-example-exists:
    description: All query parameters SHOULD have an example.
    severity: error
    message: 'Query parameter #{{value.name}} SHOULD have an example.'
    given: "$..parameters[?(@.in == 'query')]"
    then:
      field: example
      function: truthy
```

**Comment by savage-alex** [comment]
![image](https://user-images.githubusercontent.com/25430683/93359071-186fe780-f83a-11ea-9919-6839c08cdadd.png)
Looks perfect to me! (After I had remembered I had to actually build spectral and use the new binary. Such a n00b I am.

**Comment by savage-alex** [comment]
@P0lip 
I have done some more testing this morning and I think this introduced a bug:
* If you create an example property but fail to give it a value, you get an error
![image](https://user-images.githubusercontent.com/25430683/93433756-261b8080-f8bf-11ea-82a4-425784b438e0.png)

Linting the same file with the baseline finds the value missing and reports the normal error from truthy.

**Comment by P0lip** [comment]
@savage-alex could you send me the rule definition?

**Comment by savage-alex** [comment]
Sure email or have you got slack or teams so can share there?
Can give you the API def too that way. Rather than filling up GitHub in the clear.

**Comment by P0lip** [comment]
@savage-alex 
We have Stoplight Slack channel, but I'm also available on APIs You Won't Hate.
You can also send an email to jakub(at)stoplight.io.
Seems like email might be the easiest way?
